src/fortran/psychrolib.f90

```diff
@@ -403,73 +450,85 @@ function GetTDewPointFromVapPres(TDryBulb, VapPres) result(TDewPoint)
     if (isIP()) then
         BOUNDS(1) = -148.0
         BOUNDS(2) =  392.0
-        STEPSIZE  =  0.01 * 9.0 / 5.0
+        T_WATER_FREEZE = 32.
     else
         BOUNDS(1) = -100.0
         BOUNDS(2) =  200.0
-        STEPSIZE  =  0.01
+        T_WATER_FREEZE = 0.
     end if
 
-    TMidPoint = (BOUNDS(2) + BOUNDS(2)) / 2.0
-
-    ! Bounds outside which a solution cannot be found
+    ! Validity check -- bounds outside which a solution cannot be found
     if (VapPres < GetSatVapPres(BOUNDS(1)) .or. VapPres > GetSatVapPres(BOUNDS(2))) then
       error stop "Error: partial pressure of water vapor is outside range of validity of equations"
     end if
 
-    ! First guess
-    Tdp = TDryBulb
+    ! Vapor pressure contained within the discontinuity of the Pws function: return temperature of freezing
+    T_WATER_FREEZE_LOW = T_WATER_FREEZE - PSYCHROLIB_TOLERANCE / 10.        ! Temperature just below freezing
+    T_WATER_FREEZE_LOW = T_WATER_FREEZE - PSYCHROLIB_TOLERANCE / 10.        ! Temperature just below freezing
+    T_WATER_FREEZE_HIGH = T_WATER_FREEZE + PSYCHROLIB_TOLERANCE             ! Temperature just above freezing
+    PWS_FREEZE_LOW= GetSatVapPres(T_WATER_FREEZE_LOW)
+    PWS_FREEZE_HIGH = GetSatVapPres(T_WATER_FREEZE_HIGH)
+
+    ! Restrict iteration to either left or right part of the saturation vapor pressure curve
+    ! to avoid iterating back and forth across the discontinuity of the curve at the freezing point
+    ! When the partial pressure of water vapor is within the discontinuity of GetSatVapPres,
+    ! simply return the freezing point of water.
+    if (VapPres < PWS_FREEZE_LOW) then
+        BOUNDS(2) = T_WATER_FREEZE_LOW
+    else if (VapPres > PWS_FREEZE_HIGH) then
+        BOUNDS(1) = T_WATER_FREEZE_HIGH
+    else
+        TDewPoint = T_WATER_FREEZE
+        return
+    end if
+
+    ! We use NR to approximate the solution.
+    TDewPoint = TDryBulb
     lnVP = log(VapPres)
-    Tdp_c = Tdp
     index = 1
 
-    do while ((abs(Tdp - Tdp_c) > PSYCHROLIB_TOLERANCE) .or. (index < 2))
-      ! Current point
-      Tdp_c = Tdp
-      lnVP_c = log(GetSatVapPres(Tdp_c))
-      ! Step - negative in the right part of the curve, positive in the left part
-      ! to avoid going past the domain of validity of eqn. 5 and 6
-      ! when Tdp_c is close to its bounds
-      if (Tdp_c > TMidPoint) then
-        d_Tdp = -STEPSIZE
-      else
-        d_Tdp = STEPSIZE
+    do while (.true.)
+      TDewPoint_iter = TDewPoint ! TDewPoint_iter used in NR calculation
+      lnVP_iter = log(GetSatVapPres(TDewPoint_iter))
+
+      ! Derivative of function, calculated analytically
+      d_lnVP = dLnPws_(TDewPoint_iter)
+
+      ! New estimate, bounded by the search domain defined above
+      TDewPoint = TDewPoint_iter - (lnVP_iter - lnVP) / d_lnVP
+      TDewPoint = max(TDewPoint, BOUNDS(1))
+      TDewPoint = min(TDewPoint, BOUNDS(2))
+
+      if ((abs(TDewPoint - TDewPoint_iter) <= PSYCHROLIB_TOLERANCE)) then
+        return
```

Comment:
Should be:
exit
Either that or you use the do... while structure. Otherwise the last part of the function is never executed.